pkg/ruler/manager_metrics.go

```diff
@@ -99,8 +99,17 @@ func NewManagerMetrics() *ManagerMetrics {
 // AddUserRegistry adds a Prometheus registry to the struct
 func (m *ManagerMetrics) AddUserRegistry(user string, reg *prometheus.Registry) {
 	m.regsMu.Lock()
+	defer m.regsMu.Unlock()
+
 	m.regs[user] = reg
-	m.regsMu.Unlock()
+}
+
+// DeleteUserRegistry removes user-specific Prometheus registry.
+func (m *ManagerMetrics) DeleteUserRegistry(user string) {
```

Comment:
I'm not sure this is correct. The `ManagerMetrics.Collect()` sends out aggregates, so when we remove an user, the aggregated counters will decrease and will be detected as resets even if they're not. Am I missing anything?

If I'm not missing anything, I believe that what we need to do is to never remove an user registry from `ManagerMetrics` but reuse that registry if already exists.

Reply:
`Collect` method sends out all metrics *per-user*, but doesn’t aggregate over users. This is important and makes `DeleteUserRegistry` work properly. When we remove user here, given user will disappear from `Collect` output, but it doesn’t affect any aggregate values.

Now I admit this is tricky, and we should put a comment into `Collect` method to make sure we *never* do aggregates over all users there.

Reusing registry for new ruler manager wouldn’t work, since ruler manager would try to register the same metrics and fail.

We could reuse entire ruler manager, by never removing existing one from memory for given user. (But removing rules instead).

Another option is to keep a list of registries for each user. That is instead of having `map[user]Registry`, we would keep `map[user][]Registry`, and each new ruler manager for user would get its own registry. When sending metrics, we would aggregate over all of them. I think this solution would be applicable for other places in Cortex (eg. closing and reopening TSDB), but I would prefer to implement that in a separate PR, since it’s not required here (yet).

Reply:
You're right. I'm fine keeping it as is until will be required. I would just add a clear and loud comment in `Collect()` function.